File: src/Interpreter/Node.java
package Interpreter;

import java.util.ArrayList;
import java.util.function.Consumer;

public class Node {
    /**
     * Represent child and paren nodes in AST and ST
     */
    private final ArrayList<Node> children;
    private Node parent;
    private String label;
    private String value;
    /**
     *node with one argument called label
     */
    public Node(String label) {
        this.label = label;
        this.children = new ArrayList<>();
    }

    /**
     * node with two argument both lable and value
     */
    public Node(String label, String value) {
        this.label = label;
        this.value = value;
        this.children = new ArrayList<>();
    }
    Node copy() {
        Node copied = new Node(label, value);
        for (int i = 0; i < getNumChild(); i++) {
            copied.addChild(getChild(i).copy());
        }
        return copied;
    }
    //get parent node
    Node getParent() {return parent;    }
    public String getLabel() {
        return label;
    }
    public String getValue() {
        return value;
    }
    public int getNumChild() {
        return children.size();
    }
    boolean hasChildren(int n) {
        return children.size() == n;
    }
    public boolean isLabel(String label) {
        return getLabel().equals(label);
    }
    public Node getChild(int i) {
        return children.get(i);
    }
    public void forEachChild(Consumer<? super Node> action) {
        children.forEach(action);
    }
    void setLabel(String label) {this.label = label;this.value = null;    }
    void clearChildren() {children.forEach(child -> child.parent = null);children.clear();}
    void addChild(Node child) {children.add(child);child.parent = this;}
}

File: reCreate/main/INTERPRETER/Node.py
class Node:
   
    def __init__(self, label, value=None):
        self.children = []
        self.parent = None
        self.label = label
        self.value = value

    def copy(self):
        copied = Node(self.label, self.value)
        for child in self.children:
            copied.addChild(child.copy())
        return copied

    def getParent(self):
        return self.parent

    def getLabel(self):
        return self.label

    def getValue(self):
        return self.value

    def getNumChild(self):
        return len(self.children)

    def hasChildren(self, n):
        return len(self.children) == n

    def isLabel(self, label):
        return self.getLabel() == label

    def getChild(self, i):
        return self.children[i]

    def forEachChild(self, action):
        for child in self.children:
            action(child)

    def setLabel(self, label):
        self.label = label
        self.value = None

    def clearChildren(self):
        for child in self.children:
            child.parent = None
        self.children.clear()

    def addChild(self, child):
        self.children.append(child)
        child.parent = self
